File: 2023/Day_01/main.exs
defmodule Day_01 do
  def parse_input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
  end

  def part1(input) do
    input
    |> Stream.map(&String.to_charlist/1)
    |> Stream.map(fn line -> Enum.filter(line, & &1 in ?0 .. ?9) end)
    |> Stream.map(& << hd(&1), List.last(&1)>>)
    |> Stream.map(& String.to_integer/1)
    |> Enum.sum()
  end

  def part2(input) do
    input
    |> Stream.map(fn line ->
      Stream.unfold(line, fn
        "" -> nil
        line -> {line, String.slice(line, 1..-1)}
      end)
      |> Stream.map(& Regex.run(~r/\d|one|two|three|four|five|six|seven|eight|nine/, &1))
      |> Stream.filter(& &1 != nil)
      |> Stream.concat()
      |> Enum.map(fn digit ->
        Map.get(%{
          "one" => "1",
          "two" => "2",
          "three" => "3",
          "four" => "4",
          "five" => "5",
          "six" => "6",
          "seven" => "7",
          "eight" => "8",
          "nine" => "9"}, digit, digit)
        end)
      end)
    |> Stream.map(& hd(&1) <> List.last(&1))
    |> Stream.map(& String.to_integer/1)
    |> Enum.sum()
  end

  # Inspired from a Reddit comment, wrote it after finishing
  def part2v2(input) do
    numbers = "one|two|three|four|five|six|seven|eight|nine"
    first_re = Regex.compile!("\\d|#{numbers}")
    last_re = Regex.compile!("\\d|#{numbers |> String.reverse()}")
    number_mapper = fn x -> Map.get(%{
      "one" => "1",
      "two" => "2",
      "three" => "3",
      "four" => "4",
      "five" => "5",
      "six" => "6",
      "seven" => "7",
      "eight" => "8",
      "nine" => "9"}, x, x) end

    input
    |> Stream.map(fn line ->
      (Regex.run(first_re, line) |> hd() |> number_mapper.())
      <> (Regex.run(last_re, line |> String.reverse()) |> hd() |> String.reverse() |> number_mapper.())
    end)
    |> Stream.map(& String.to_integer/1)
    |> Enum.sum()
  end

  def part2v3(input) do
    input
    |> Enum.map(fn line ->
      line
      |> String.replace("one", "o1e")
      |> String.replace("two", "t2o")
      |> String.replace("three", "th3ee")
      |> String.replace("four", "f4ur")
      |> String.replace("five", "f5ve")
      |> String.replace("six", "s6x")
      |> String.replace("seven", "se7en")
      |> String.replace("eight", "ei8ht")
      |> String.replace("nine", "ni9e")
      end)
    |> part1()
  end

  def main do
    filename = Enum.at(System.argv, 0, "input.txt")
    parsed_input = parse_input(filename)

    IO.inspect(part1(parsed_input), label: "Part 1")
    IO.inspect(part2v3(parsed_input), label: "Part 2")
  end
end

Day_01.main()

File: 2023/Day_01/main.hs
import Data.Function ((&))
import Data.Functor ((<&>))
import Data.Char (isDigit, digitToInt, intToDigit)
import Data.List (tails)
import Data.Maybe (fromJust)

part1 =
    sum
    . map (read
        . (\l -> head l : last l : "")
        . filter isDigit)

getNumber "" = Nothing
getNumber (x:xs)
    | isDigit x = Just $ digitToInt x
    | otherwise = case x:xs of
        'o':'n':'e':_ -> Just 1
        't':'w':'o':_ -> Just 2
        't':'h':'r':'e':'e':_ -> Just 3
        'f':'o':'u':'r':_ -> Just 4
        'f':'i':'v':'e':_ -> Just 5
        's':'i':'x':_ -> Just 6
        's':'e':'v':'e':'n':_ -> Just 7
        'e':'i':'g':'h':'t':_ -> Just 8
        'n':'i':'n':'e':_ -> Just 9
        _ -> getNumber xs

getLastNumber = head . dropWhile (== Nothing) . map getNumber . reverse . tails

part2 =
    sum
    . map (read . (\line -> intToDigit (fromJust (getNumber line)) : intToDigit (fromJust (getLastNumber line)) : ""))

main = do
    input <- readFile "input.txt" <&> lines
    let ans1 = part1 input
    putStrLn $ "Part 1: " ++ show ans1
    let ans2 = part2 input
    putStrLn $ "Part 2: " ++ show ans2